include/dlpack/dlpack.h

```diff
@@ -200,27 +210,60 @@ typedef struct {
   int64_t* strides;
   /*! \brief The offset in bytes to the beginning pointer to data */
   uint64_t byte_offset;
-} DLTensor;
+  /*! \brief The DLPack and DLPack ABI versions. The exporting library
+   *  should set the DLPack version to DLPACK_VERSION and ABI version to
+   *  DLPACK_ABI_VERSION. */
+  DLPackVersion version;
+  /*! \brief Mark the data readonly. */
+  uint8_t readonly;
+} DLTensorVersioned;
```

Comment:
Actually one thing that might be useful is to think about how to minimizing the change. Specifically, in light of S0 in https://github.com/dmlc/dlpack/issues/104. We may not need to introduce a separate versioned struct. 

The old framework can still use the DLTensor(as implemented in the versioned version) as it is, as long as there is no reliance of sizeof(DLTensor). Perhaps we can rename the old one as DLTensorLegacy; In the similar spirit, it is safe to `static_cast`  a DLTensor* to DLTensorLegacy*, as a result, we do not need to introduce two versions of DLManagedTensor

Reply:
> The old framework can still use the DLTensor(as implemented in the versioned version) as it is, as long as there is no reliance of sizeof(DLTensor). Perhaps we can rename the old one as DLTensorLegacy; In the similar spirit, it is safe to `static_cast` a DLTensor* to DLTensorLegacy*, as a result, we do not need to introduce two versions of DLManagedTensor

But DLManagedTensor doesn't have a pointer to the DLTensor struct. Instead, it is just a plain `DLTensor dl_tensor` field. So, changing the sizeof(DLTensor) will change alignment and size of DLManagedTensor. Which is why we need to rename both structs. I also considered renaming the old structs to `DLTensorLegacy` and `DLManagedTensorLegacy` but it seems equivalent to instead add new structs with a different name (new implementations will be able to see both and can use the struct they want to export...)

Reply:
Ah sorry, you are indeed right. This would make the change ABI breaking. Although it would make future ABI change hard because all of them will be non-backward compact due to the change of deleter offset. Would be great to have more deliberation of this topic.

One possible choice is to append other flags(version, read_only) to DLManagedTensor instead. This of course won't be too ideal in case we want to change fields in DLTensor, but would maintain backward compact

Reply:
Ahh, I see. If we move them to DLManagedTensor, we can ensure the location of the deleter function is still at the same offset. Then the code 
```
DLManagedTensorVersioned *managed = (DLManagedTensorVersioned *)PyCapsule_GetPointer(self, "dltensor");
if (managed->deleter) {
         managed->deleter(managed);
```

Reply:
Although not ideal, I too think it'd be better to add the new fields to the DLManagedTensor instead. This would eliminate the requirement of the `*Versioned` structs and also make it easier to implement support for the new ABI.